docs/content/sips/000-spin-test.md

```diff
@@ -0,0 +1,189 @@
+    
+    /// The virtualization needs the component's id and
+    /// the `spin.toml` manifest because many of the spin
+    /// interfaces are configured through that combination.
+    export set-component-id: func(component-id: string);
+    import get-manifest: func() -> string;
+
+	/// A way to say that a certain URL is associated with a response
+    export http-handler;
+    /// How the calls to the kv store are tracked
+    export key-value-calls;
+    
+}
+
+interface http-handler {
+    use wasi:http/types@0.2.0.{future-incoming-response};
+    set-response: func(url: string, response: future-incoming-response);
+}
+
+interface key-value-calls {
+    reset: func();
+    get: func() -> list<tuple<string, list<get-call>>>;
+    set: func() -> list<tuple<string, list<set-call>>>;
+
+    record get-call {
+        key: string
+    }
+
+    record set-call {
+        key: string,
+        value: list<u8>
+    }
+}
+```
+
+### The `fermyoin:spin-test/runner` world
+
+```
+world runner {
+	/// The host must supply a manifest
+    import get-manifest: func() -> string;
+    /// The host must give the runner the ability to create HTTP resources
+    import http-helper;
+    /// The host can then run the test
+    export run: func();
+}
+```
+
+## Open Questions
+
+- Should `spin-test` be a stand alone CLI experience, a plugin for `spin` CLI, or both?
+- How do we handle Spin applications with more than component?
+    - There’s no way to compose a statically defined component with N number of components. This means the router component must know how many components it will be composed with ahead of time, and so the reasonable number of 1 is picked.
+    - We could work around this by moving the router out of a component and into the host, but this makes the host smarter which would like to avoid. We could also potentially synthesize a router component based on the `spin.toml` manifest instead of relying on a statically defined one, but this is certainly not trivial.
+- How should test discovery work?
+    - The most straightforward way would be for `spin-test` to require the user to specify the path to a `spin-test` compliant test component binary.
+    - Another way would be for there to be a convention around a `tests` directory. This `tests` directory could include both built `spin-test` compliant test component binaries as well as directories with source code and a `spin-test.toml` configuration file that specifies how the source code is built into a `spin-test` compliant test component binary. `spin-test` would then build all the tests that need to be built and run each test binary.
+- How exactly should the various worlds look like?
+    - The proposal above is enough to get things working, but there is plenty of room for bike-shedding the exact shape of the worlds.
+- Which triggers should we support?
+    - Can we support an arbitrary number of triggers through some sort of plugin system? How would that work?
```

Comment:
This feels like something that may become dramatically easier with CM async (to the extent that magically fixes all our nested composition woes).

Reply:
Hmm... I'm not sure I follow - none of the triggers (with the exception of `wasi:http/incoming-handler` which we already support) are necessarily built to try to accommodate async. I don't see how CM async improves the situation here. Could you explain more about what you mean?

Reply:
Ah yeah sorry the train of consciousness left the station early there...

I think each trigger type could have a test driver adapter to wrap the component-under-test and test driver to implement the runner world; this is impossible today because the host has to provide resource factories to glue everything together.